File: Aigilas/Aigilas/HUD/InventoryHud.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Aigilas.Creatures;
using Aigilas.Items;
using SPX.Util;
using SPX.Core;
using Aigilas.Management;
using SPX.IO;

namespace Aigilas.HUD
{
    public class InventoryHud:IHud
    {
        private int _currentClass = ItemClass.Values[1];
        private readonly Inventory _inventory;
        private readonly Equipment _equipment;
        private int _endingItem = 4;
        private int _startingItem = 0;
        private Dictionary<GenericItem, int> _currentClassItems;
        private GenericItem _currentSelectedItem = null;

        private EquipmentHud _equipHud;
        private DeltasHud _deltas;

        public InventoryHud(ICreature owner,Inventory inventory,Equipment equipment):base(owner,XnaManager.WindowWidth/2,XnaManager.WindowHeight/2)
        {
            _inventory = inventory;
            _equipment = equipment;
            _deltas = new DeltasHud(owner,equipment);
            _equipHud = new EquipmentHud(owner, equipment);
            _currentClassItems = _inventory.GetItems(_currentClass);
        }

        public void Draw()
        {
            if (_isVisible)
            {
                XnaManager.Renderer.Draw(_menuBase,GetHudOrigin(), new Rectangle(0, 0, 1, 1), Color.Black, 0f, new Vector2(0,0), XnaManager.GetCenter(), SpriteEffects.None,ZDepth.HudBG);
                _textHandler.Draw();
                _deltas.Draw();
                _equipHud.Draw();
            }
        }

        private void HandleInput()
        {
            if (Input.IsActive(Commands.CycleLeft, _parent.GetPlayerIndex()))
            {
                _currentClass--;
                if (_currentClass <= ItemClass.NULL)
                {
                    _currentClass = ItemClass.Values.Count() - 2;
                }
                _startingItem = 0;
                _endingItem = 4;
                forceRefresh = true;
            }

            if (Input.IsActive(Commands.CycleRight, _parent.GetPlayerIndex()))
            {
                _currentClass++;
                if (_currentClass >= ItemClass.LAST)
                {
                    _currentClass = ItemClass.Values[1];                    
                }
                _startingItem = 0;
                _endingItem = 4;
                forceRefresh = true;
            }

            if (Input.IsActive(Commands.MoveDown, _parent.GetPlayerIndex()))
            {
                if (_startingItem < _currentClassItems.Count() - 1)
                {
                    _startingItem++;
                    _endingItem++;
                    forceRefresh = true;
                }
            }

            if (Input.IsActive(Commands.MoveUp, _parent.GetPlayerIndex()))
            {
                if (_startingItem > 0)
                {
                    _startingItem--;
                    _endingItem--;
                    forceRefresh = true;
                }
            }
            if (Input.IsActive(Commands.Confirm, _parent.GetPlayerIndex()))
            {
                _parent.Equip(_currentSelectedItem);
                forceRefresh = true;
            }
            if (Input.IsActive(Commands.Cancel, _parent.GetPlayerIndex()))
            {
                _parent.Drop(_currentSelectedItem);
                forceRefresh = true;
            }
        }
        public void Update()
        {
            if (_isVisible)
            {
                HandleInput();
                _textHandler.Update();
                _deltas.Update(_currentSelectedItem,forceRefresh);
                _equipHud.Update(forceRefresh);
                _textHandler.Clear();
                UpdateInventoryDisplay();
                if (forceRefresh)
                {
                    forceRefresh = false;
                }
            }
        }

        public override void Toggle()
        {
            base.Toggle();
            _deltas.Toggle();
            _equipHud.Toggle();
            forceRefresh = true;
        }

        private static readonly Dictionary<int, string> __classStrings = new Dictionary<int, string>();

        private string GetClassDisplay()
        {
            if(!__classStrings.ContainsKey(_currentClass))
            {
                __classStrings.Add(_currentClass, ItemClass.Names[_currentClass]);
            }
            return __classStrings[_currentClass];
        }

        private const string __delim = ")";
        private const string __equipDelim = "~";
        private const string __seper = " x";
        private const string __newline = "\n";

        private string displayString = "";
        private bool forceRefresh = false;

        private void UpdateInventoryDisplay()
        {
            _textHandler.WriteDefault(GetClassDisplay(), 20, 30,GetHudOrigin());
            _currentClassItems = _inventory.GetItems(_currentClass);
            if (_currentClassItems.Count > 0)
            {
                int ii = 0;
                if (forceRefresh)
                {
                    StringSquisher.Clear();
                    foreach (var item in _currentClassItems.Keys)
                    {
                        if (ii == _startingItem)
                        {
                            _currentSelectedItem = item;
                        }
                        if(!_equipment.IsRegistered(item)&&_inventory.GetItemCount(item)<=0)
                        {
                            continue;
                        }

                   
                        if (ii >= _startingItem && ii < _endingItem && ii < _currentClassItems.Keys.Count())
                        {
                            StringSquisher.Squish
                            (
                                StringStorage.Get(ii), __delim,
                                (_equipment.IsRegistered(item)) ? __equipDelim : String.Empty,
                                item.Name);
                            if (_currentClassItems[item] > -1)
                            {
                                StringSquisher.Squish(__seper, StringStorage.Get(_currentClassItems[item]));
                            }
                            StringSquisher.Squish(__newline);
                        }
                        ii++;
                    }
                    displayString = StringSquisher.Flush();       
                }
                _textHandler.WriteDefault(displayString, 50, 60, GetHudOrigin());
            }
        }
    }
}

File: aigilas/src/aigilas/hud/InventoryHud.java
package aigilas.hud;

import aigilas.Common;
import aigilas.creatures.BaseCreature;
import aigilas.items.Equipment;
import aigilas.items.GenericItem;
import aigilas.items.Inventory;
import aigilas.items.ItemClass;
import com.badlogic.gdx.graphics.Color;
import sps.bridge.Commands;
import sps.bridge.DrawDepths;
import sps.core.Point2;
import sps.graphics.Renderer;
import sps.io.Input;
import sps.text.Text;
import sps.text.TextPool;
import sps.util.StringSquisher;
import sps.util.StringStorage;

import java.util.HashMap;

public class InventoryHud extends BaseHud {
    private int _currentClass = 0;
    private final Inventory _inventory;
    private final Equipment _equipment;
    private int _endingItem = 4;
    private int _startingItem = 0;
    private HashMap<GenericItem, Integer> _currentClassItems;
    private GenericItem _currentSelectedItem = null;

    private final EquipmentHud _equipHud;
    private final DeltasHud _deltas;

    public InventoryHud(BaseCreature owner, Inventory inventory, Equipment equipment) {
        super(owner, Renderer.VirtualWidth / 2, Renderer.VirtualHeight / 2);
        _inventory = inventory;
        _equipment = equipment;
        _deltas = new DeltasHud(owner, equipment);
        _equipHud = new EquipmentHud(owner, equipment);
        _currentClassItems = _inventory.getItems(ItemClass.values()[_currentClass]);
    }

    @Override
    public void draw() {
        if (_isVisible) {
            Renderer.get().draw(_menuBase, getInventoryAnchor(), DrawDepths.get(Common.Hud_BG), Color.BLACK, (int) Renderer.get().center().X, (int) Renderer.get().center().Y);
        }
    }

    private void handleInput() {
        if (Input.isActive(Commands.get(Common.CycleLeft), _parent.getPlayerIndex())) {
            _currentClass--;
            if (_currentClass < 0) {
                _currentClass = ItemClass.values().length - 1;
            }
            _startingItem = 0;
            _endingItem = 4;
            forceRefresh = true;
        }

        if (Input.isActive(Commands.get(Common.CycleRight), _parent.getPlayerIndex())) {
            _currentClass++;
            if (_currentClass >= ItemClass.values().length) {
                _currentClass = 0;
            }
            _startingItem = 0;
            _endingItem = 4;
            forceRefresh = true;
        }

        if (Input.isActive(Commands.get(Common.MoveDown), _parent.getPlayerIndex())) {
            if (_startingItem < _currentClassItems.size() - 1) {
                _startingItem++;
                _endingItem++;
                forceRefresh = true;
            }
        }

        if (Input.isActive(Commands.get(Common.MoveUp), _parent.getPlayerIndex())) {
            if (_startingItem > 0) {
                _startingItem--;
                _endingItem--;
                forceRefresh = true;
            }
        }
        if (Input.isActive(Commands.get(Common.Confirm), _parent.getPlayerIndex())) {
            _parent.equip(_currentSelectedItem);
            forceRefresh = true;
        }
        if (Input.isActive(Commands.get(Common.Cancel), _parent.getPlayerIndex())) {
            _parent.drop(_currentSelectedItem);
            forceRefresh = true;
        }
    }

    public void update() {
        if (_isVisible) {
            handleInput();
            updateInventoryDisplay();
        }
        else {
            if (header != null) {
                header.hide();
                itemClass.hide();
                for (Text text : itemList) {
                    if (text != null) {
                        text.hide();
                    }
                }
            }
        }
        _equipHud.update(forceRefresh);
        _deltas.update(_currentSelectedItem, forceRefresh);
        if (forceRefresh) {
            forceRefresh = false;
        }
    }

    @Override
    public void toggle() {
        super.toggle();
        _deltas.toggle();
        _equipHud.toggle();
        forceRefresh = true;
    }

    private static final HashMap<Integer, String> __classStrings = new HashMap<Integer, String>();

    private String getClassDisplay() {
        if (!__classStrings.containsKey(_currentClass)) {
            __classStrings.put(_currentClass, ItemClass.values()[_currentClass].hudName());
        }
        return __classStrings.get(_currentClass);
    }

    private static final String __delim = ")";
    private static final String __equipDelim = "~";
    private static final String __seper = " x";

    private String displayString = "";
    private boolean forceRefresh = false;
    private String[] list = new String[10];

    private Text header;
    private Text itemClass;
    private Text[] itemList = new Text[10];

    private Point2 inventoryPosition = new Point2(0, 0);

    private void updateInventoryDisplay() {
        //This needs to be rendered up here, because it won't be rendered if the class contains no items
        if (forceRefresh) {
            if (header != null) {
                header.hide();
                itemClass.hide();
                for (int ii = 0; ii < 10; ii++) {
                    if (itemList[ii] != null) {
                        itemList[ii].hide();
                    }
                }
            }
            inventoryPosition.reset(getInventoryAnchor().X + 50, getInventoryAnchor().Y + (int) (_dimensions.Y * .9), false);
            header = TextPool.get().write(getClassDisplay(), inventoryPosition);


            list = new String[10];

            _currentClassItems = _inventory.getItems(ItemClass.values()[_currentClass]);
            if (_currentClassItems.size() > 0) {
                int ii = 0;

                StringSquisher.clear();
                displayString = StringSquisher.flush();
                int count = 0;
                for (GenericItem item : _currentClassItems.keySet()) {
                    if (ii == _startingItem) {
                        _currentSelectedItem = item;
                    }
                    if (!_equipment.isRegistered(item) && _inventory.getItemCount(item) <= 0) {
                        continue;
                    }

                    if (ii >= _startingItem && ii < _endingItem && ii < _currentClassItems.keySet().size()) {
                        count++;
                        StringSquisher.clear();
                        StringSquisher.squish(StringStorage.get(ii), __delim, (_equipment.isRegistered(item)) ? __equipDelim : "", item.Name);
                        if (_currentClassItems.get(item) > -1) {
                            StringSquisher.squish(__seper, StringStorage.get(_currentClassItems.get(item)));
                        }
                        if (count < 10) {
                            list[count] = StringSquisher.flush();
                        }
                    }
                    ii++;
                }

            }
            inventoryPosition.reset(getInventoryAnchor().X + 50, getInventoryAnchor().Y + 60, false);
            if (itemClass != null) {
                itemClass.hide();
            }
            itemClass = TextPool.get().write(displayString, inventoryPosition);
            for (int jj = 0; jj < 10; jj++) {
                if (itemList[jj] != null) {
                    itemList[jj].hide();
                }
                if (list[jj] != null) {
                    inventoryPosition.reset(getInventoryAnchor().X + 40, getInventoryAnchor().Y + (int) (_dimensions.Y * .9) - 60 * jj, false);
                    itemList[jj] = TextPool.get().write(list[jj], inventoryPosition);
                }
            }
        }
    }
}
